lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -3664,10 +3654,10 @@ ConformanceChecker::inferTypeWitnessesViaValueWitness(ValueDecl *req,
   }
 
   auto setup = [&]() -> std::tuple<Optional<RequirementMatch>, Type, Type> {
-    fullWitnessType = removeSelfParam(witness, fullWitnessType);
+    fullWitnessType = fullWitnessType->removeSelfParam(witness);
     return std::make_tuple(
         None,
-        removeSelfParam(req, req->getInterfaceType()),
+        req->getInterfaceType()->removeSelfParam(req),
```

Comment:
Also is req and witness always in type context? Then you could have an assert here, instead of an 'if'.